spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala

```diff
@@ -429,10 +429,46 @@ trait Checkpoints extends DeltaLogging {
    */
   private[delta] def findLastCompleteCheckpointBefore(
       checkpointInstance: Option[CheckpointInstance] = None): Option[CheckpointInstance] = {
-    val upperBoundCv = checkpointInstance.filterNot(_.version < 0).getOrElse {
-        logInfo(s"Try to find Delta last complete checkpoint")
-        return findLastCompleteCheckpoint()
-      }
+    val eventData = mutable.Map[String, String]()
+    val startTimeMs = System.currentTimeMillis()
+    def sendUsageLog(): Unit = {
+      eventData("totalTimeTakenMs") = (System.currentTimeMillis() - startTimeMs).toString
+      recordDeltaEvent(
+        self, opType = "delta.findLastCompleteCheckpointBefore", data = eventData.toMap)
+    }
+    try {
+      val resultOpt = findLastCompleteCheckpointBeforeInternal(eventData, checkpointInstance)
+      eventData("resultantCheckpointVersion") = resultOpt.map(_.version).getOrElse(-1L).toString
+      sendUsageLog()
+      resultOpt
+    } catch {
+      case e@(NonFatal(_) | _: InterruptedException | _: java.io.InterruptedIOException |
+              _: java.nio.channels.ClosedByInterruptException) =>
+        eventData("exception") = Utils.exceptionString(e)
+        sendUsageLog()
+        throw e
+    }
+  }
+
+  private def findLastCompleteCheckpointBeforeInternal(
+      eventData: mutable.Map[String, String],
+      checkpointInstance: Option[CheckpointInstance]): Option[CheckpointInstance] = {
+    val upperBoundCv =
+      checkpointInstance
+        // If someone passes the upperBound as 0 or sentinel value, we should not do backward
+        // listing. Instead we should list the entire directory from 0 and return the latest
+        // available checkpoint.
+        .filterNot(cv => cv.version < 0 || cv.version == CheckpointInstance.MaxValue.version)
```

Comment:
This is the fix and there is a corresponding test: `findLastCompleteCheckpoint with CheckpointInstance.MAX value` .